src/Microsoft.Health.Fhir.Liquid.Converter/Utilities/ComplexObjectFilterUtility.cs

```diff
@@ -0,0 +1,173 @@
+            return ret.ToArray<object>();
+        }
+
+        public static bool ObjHasValueAtPath(object input, Queue<string> path, string value)
+        {
+            // If we're at the end of the path then check the value
+            if (path.Count == 0)
+            {
+                // Return true if value is null and an equality check otherwise
+                return value == null || input.ToString().Equals(value);
+            }
+
+            // Get our key name
+            var key = path.Dequeue();
+
+            // Check if our key is an object property or an array reference
+            if (key.Contains('['))
+            {
+                // Our required path dictates we should have an array and if not
+                // then this is not a match
+                if (input is object[] inputArray)
+                {
+                    // If key is [] then we want to loop over every object in the array
+                    // and check for our future paths
+                    if (key == "[]")
+                    {
+                        if (path.Count == 0)
+                        {
+                            return (value == null) || inputArray.Contains(value);
+                        }
+
+                        foreach (object obj in inputArray)
+                        {
+                            // Clone the queue to scope the path at this level for each loop
+                            var localPath = new Queue<string>(path);
+                            if (ObjHasValueAtPath(obj, localPath, value))
+                            {
+                                return true;
+                            }
+                        }
+
+                        return false;
+                    }
+
+                    // If our array key references a specific index
+                    else
+                    {
+                        // Trim and parse out our numeric index
+                        var indexStr = key.TrimStart('[').TrimEnd(']');
+                        var index = int.Parse(indexStr);
+
+                        if (index >= 0 && inputArray.Count() > index)
+                        {
+                            // Recurse
+                            return ObjHasValueAtPath(inputArray[index], path, value);
+                        }
+                        else
+                        {
+                            // It's not possible for this array to have the element specified
+                            // so this is not a match
+                            return false;
+                        }
+                    }
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            // Our key is referencing an object property
+            else
+            {
+                if (input is Dictionary<string, object> inputObject)
+                {
+                    if (inputObject.ContainsKey(key))
```

Comment:
For readability I would combine the two if statements with an `&&` instead of having two separate blocks. You could also just return the value comparison instead of explicitly returning a false value.

```csharp
return input is Dictionary<string, object> inputObject 
  && inputObject.ContainsKey(key) 
  && ObjHasValueAtPath(inputObject[key], path, value);
```

Reply:
On one hand I like condensing the 15 lines to 3 to make it clear that the return value will be based on those 3 conditions instead of all the potential branches from the ifs.  On the other hand, I like the else on 106 because it makes it more clear that this will only happen if `if (key.Contains('['))` is false - even though in reality all the branches are closed with returns above it.

Reply:
Thank you!